### PR TITLE
Optional weights_square_array parameter for UVFlag

### PR DESCRIPTION
**Pull request by mwilensky768**
This change adds an optional "weights_square_array" parameter to UVFlag. This is only calculated in `UVFlag.to_waterfall` if the object starts as `baseline` type. This parameter holds the sum of the squares of the weights along the collapsed (baseline) axis. 

## Description

This adds an optional parameter that is set up during `__init__` and a couple other places (see code changes). This initial setup of the parameter currently needs more work (see comments below). This adds a keyword to `UVFlag.to_waterfall` called `return_weights_square`. When this is `True`, `uvutils.collapse` is called with `return_weights_square=True`, and the attribute `UVFlag.weights_square_array` is set to the output from the collapse function. 

## Motivation and Context

This lets the user obtain the sum of the squares of the weights over the baselines when making a waterfall object, which is useful when using general weights.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Bug fix checklist:
- [ ] My fix includes a new test that breaks as a result of the bug (if possible).
- [ ] All new and existing tests pass.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Breaking change checklist:
- [ ] I have updated the docstrings associated with my change using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to reflect my changes (if appropriate).
- [ ] My change includes backwards compatibility and deprecation warnings (if possible).
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests pass.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Documentation change checklist:
- [ ] Any updated docstrings use the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If this is a significant change to the readme or other docs, I have checked that they are rendered properly on ReadTheDocs. (you may need help to get this branch to build on RTD, just ask!)

Version change checklist:
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md) to put all the unreleased changes under the new version (leaving the unreleased section empty).
- [ ] I have noted any dependency changes since the last version and will update the conda package build accordingly.

Build or continuous integration change checklist:
- [ ] If required or optional dependencies have changed (including version numbers), I have updated the readme to reflect this.
- [ ] If this is a new CI setup, I have added the associated badge to the readme and to references/make_index.py (if appropriate).

### CODE REVIEW

**Comment by mwilensky768** [comment]
@bhazelton @mkolopanis there is something very silly happening. Given the changes that I have made, an error pops up during `UVFlag.clear_unused_attributes` where it complains about not having an attribute named `N`. If I do not have line `742` in `uvflag.py`, this problem doesn't happen, except then the new attribute (`weights_square_array`) gets set to `None`. I feel that making it required just for this instance is bad practice. Please help.

**Comment by bhazelton** [comment]
@mwilensky768 I found and fixed a couple of typos that I think were causing the problems you saw.

**Comment by mwilensky768** [comment]
@bhazelton you have eagle eyes! Thank you.

**Comment by mwilensky768** [comment]
So, it seems all of what I've added is covered, but I'm not sure it's sufficient to fully incorporate the array as an optional parameter. Is there a type of test that can be performed to see if a parameter has all the necessary bells and whistles?

**Comment by bhazelton** [comment]
@mwilensky768 you should check that if you write the object out and then read it back in it's identical -- that checks that your new parameter is round tripped through file write/read.

File: pyuvdata/uvflag/uvflag.py [hunk 1]
```diff
@@ -667,6 +682,7 @@ def _set_type_waterfall(self):
         self._Nants_data.required = False
         self._Nbls.required = False
         self._Nspws.required = False
+        self._weights_square_array.required = False
```
Comment (bhazelton) [inline pyuvdata/uvflag/uvflag.py]: I'm not sure this line is necessary. It's initialized as not required and I don't see anywhere that `_weights_square_array.required` is ever set to True.

Comment (mwilensky768) [reply pyuvdata/uvflag/uvflag.py]: Oops! I meant to remove this one. I'll remove it in my next push after I write a round trip test.

**Comment by mwilensky768** [comment]
OK, a round-trip test has been added and it succeeds.

**Comment by bhazelton** [comment]
I think it looks good too as soon as the listing issue is fixed.

**Comment by bhazelton** [comment]
It probably merits an update to the changelog as well. I don't know if the tutorial needs anything added.

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 - `frequency_average` method on UVData to average data along the frequency axis.
+- `weights_square_array` (optional) parameter on UVFlag - stores sum of squares of weights when converting to waterfall
```
Comment (bhazelton) [inline CHANGELOG.md]: this should be at the top -- the lines should be in reverse chronological order

**Comment by mwilensky768** [comment]
Fixed it.

**Comment by mkolopanis** [comment]
Thanks for the changelog changes. looks like a rebase/conflict correction is needed then you should be good I think.

**Comment by mwilensky768** [comment]
@mkolopanis I fixed the merge conflict (I think it was a whitespace error or something). It looks like it's hung up because the python 3.8 env won't build properly on macos.

**Comment by bhazelton** [comment]
I'm not worried about mac 3.8 builds, they're flaky. I was waiting for the codecov report but it seems stuck. I will give it a poke.

**Pull request merged**